### ISSUE TITLE
GHC consistency between hsdev and user's codebase

### ISSUE THREAD

**Issue by bscottm** [SublimeHaskell/SublimeHaskell#318]
This will likely become a Frequently Reported Bug: The GHC used to compile `hsdev` needs to match the GHC that the user's development GHC. This afflicts developers who have multiple GHC installs or use `stack` with a project-specific GHC or multiple LTSes.

Problem: How to find the "correct" `hsdev`?

**Comment by iamrecursion** [comment]
I was thinking about this today actually. Once the current backend refactoring is complete do you think it would be feasible to have an option to look in the local stack binaries for `hsdev`? At least that way the user has the option of running `stack install hsdev` for the Stackage LTS they're running. Small issue being that `hsdev` has had issues building on certain stack LTS releases in the past (though that seems to be cleared up somewhat these days). 

It's something I'd be willing to try my hand at hacking on once my current deadlines are past.

**Comment by bscottm** [comment]
@iamrecursion: Yes, it would be very possible to add a `hsdev` backend option to `stack exec -- hsdev`. It would also be possible to `cabal run` it too.

I think you'd want to `stack build hsdev` in the current project directory, then use `stack exec -- hsdev` to execute it. That's the recommended way for using `intero`.

**Comment by iamrecursion** [comment]
That was pretty much along the lines of what I was thinking. If you're not up for looking into this I'm happy to take it onto my plate once the current back-end refactor is stable. Goodness knows it's about time for me to give back to a plugin that's given me so much over the past year. 

In my mind the ideal behaviour would be to act as follows for stack sandboxes (though I'd imagine something similar could work for cabal):
1. Speculatively execute `stack exec -- hsdev <opts>`
2. In case of error, provide message to user and fall back to `hsdev` on system path (with standard error behaviour should this not exist). The ideal message would warn about possible discrepancies between `hsdev` and `ghc` output. 

It might also be possible to provide an 'automatic' install option that builds hsdev in the local sandbox for the user, starting it automatically once complete. If this were the case it would definitely be possible to have a preference for `auto_install_hsdev_in_stack_sandbox` (or similar) to save that manual step for every project. That way you could guarantee that output from `hsdev` and `ghc` match as closely as possible (and it might also avoid some of the stranger issues I've had with `hsdev` being unable to locate libraries). 

Any thoughts would be appreciated. I've not done a _huge_ amount of poking around in either the `SublimeHaskell` or `hsdev` codebases yet, so I don't know if that proposed behaviour would introduce issues.

**Comment by bscottm** [comment]
@iamrecursion: The main "problem" is blocking the main Python thread, which also blocks ST's user interface from reacting to anything. So, even though something useful is happening, ST looks like it has locked up. I haven't figured out a good way to defer certain operations (completion queries, for example) until the backend is actually up and running. I suppose that returning no completions if there's no backend is an option -- that would give the sandbox build time to complete.

I wouldn't make `auto_install_hsdev_in_stack_sandbox` a preference. It can now be a backend-specific option:

```
    "backends": {
        "Remote hsdev@localhost:4567": {
            "default": true,
            "backend": "hsdev",
            "options": {
                "local": true,
                "exec_in_sandbox": true,
                "build_if_needed": true
            }
        }
    }
```

**Comment by iamrecursion** [comment]
That's a much nicer solution for the options than what I was thinking. It sounds like the optimal solution is then to automatically swap to 'no' backend while the build is taking place and then automatically swap back. I don't see why this would be too much of an issue as long as the user was provided with adequate feedback. 

I'll have a look at this once the backend refactor is into stable if you'd like.

**Comment by bscottm** [comment]
@iamrecursion: Thanks for the offer! I'll definitely take you up on it!

**Comment by iamrecursion** [comment]
So I've been looking into this more and it seems that most stackage snapshots don't leave hsdev in a state where it can be built. The current `nightly`, for example, has dependencies too new:
```
In the dependencies for hsdev-0.2.3.2:
    hformat-0.3.0.0 must match ==0.2.* (latest applicable is 0.2.0.0)
    simple-log-0.9.2 must match ==0.8.* (latest applicable is 0.8.5)
    text-region-0.3.0.0 must match ==0.2.* (latest applicable is 0.2.0.0)
```
The current `lts-8.15`, runs into similar issues, recommending to add explicit `extra-deps` to get around it, which doesn't fix the issue as the `lts` release has dependencies too old for other things. This means that, even in the latest `lts` release, there would be no way to build a `hsdev` executable via `stack` that would sync with that project's `ghc` version.
```
Error: While constructing the build plan, the following exceptions were encountered:

In the dependencies for hsdev-0.2.3.2:
    haddock-api must match >=2.17.0 && <2.18.0, but the stack configuration has no specified version
                (latest applicable is 2.17.4)
    hdocs must match >=0.5.0, but the stack configuration has no specified version (latest applicable is 0.5.2.0)
    hformat-0.1.0.1 must match ==0.2.* (latest applicable is 0.2.0.0)
    simple-log-0.5.1 must match ==0.8.* (latest applicable is 0.8.5)
    text-region-0.1.0.1 must match ==0.2.* (latest applicable is 0.2.0.0)
```

If we want to get this working, there'd have to be some fairly drastic changes somewhere. One possible solution would be to keep the version boundaries for `hsdev` up to date with `nightly`, but I imagine that even with someone actively maintaining that it could get ugly quite quickly.

Nevertheless, the mismatch between `hsdev` and `ghc` outputs is a concerning issue and I'm wondering if there isn't some other way to deal with this that wouldn't be so prone to breakage. 

One (_very_ ugly) hack that I've managed to make work (manually) is to `stack install cabal-install` and then install `hsdev` inside a cabal sandbox in the project directory (so as to avoid polluting any system-level things). It's horrible, but it _does_ work. Nevertheless, I think keeping the dependencies of `hsdev` more up to date might still be a better solution than something that is, by it's inherent nature, a bodge. 

One (annoying) result of this situation is that there is no way to have a conclusively working set of the tools required by `SublimeHaskell` globally _without_ the use of cabal.

**Comment by iamrecursion** [comment]
`hsdev` is now submitted to stackage, courtesy of @mvoidex: fpco/stackage/pull/2545. Once it makes it into nightly this means that we can put together a reasonable plan for getting this working in preparation for the backend refactor landing.

**Comment by bscottm** [comment]
Actually, it wasn't that hard to add the ability to `stack exec` or `cabal exec` to the `hsdev` backend. `gcc-mod` can't do it because it has to execute in the same directory as the project cabal it's operating on.

**Comment by iamrecursion** [comment]
So where does that leave this work? Still need to add the ability to automatically install it in the sandbox?

**Comment by bscottm** [comment]
@iamrecursion: Yes, the ability to automagically install in the sandbox is still an missing feature. I ended up needing to test `hsdev` in a sandboxed environment, which precipitated the hack.

**Comment by iamrecursion** [comment]
Cool. Once the backend refactor is stable I guess we can work out exactly how automatic installation should behave.
